pip/pip-278.md

```diff
@@ -0,0 +1,131 @@
+
+Class Diagram of core class:
+![pip-278-1.png](images/pip-278-1.png)
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* Define a standard TopicCompactionService interface.
+
+  ```java
+  public interface TopicCompactionService {
+      /**
+       * Compact the topic.
+       * Topic Compaction is a key-based retention mechanism. it will keep the most recent value for a given key and
+       * user will read compacted data from TopicCompactionService.
+       *
+       * @return a future that will be completed when the compaction is done.
+       */
+      CompletableFuture<Void> compact();
+
+      /**
+       * Read the compacted entries from the TopicCompactionService.
+       *
+       * @param startPosition         the position to start reading from.
+       * @param numberOfEntriesToRead the number of entries to read.
+       * @return a future that will be completed with the list of entries, this list can is null.
+       */
+      CompletableFuture<List<Entry>> readCompactedEntries(@NonNUll Position startPosition, int numberOfEntriesToRead);
+
+      /**
+       * Read the last compacted entry from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the compacted last entry, this entry can is null.
+       */
+      CompletableFuture<Entry> readCompactedLastEntry();
+
+      /**
+       * Get the last compacted position from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the last compacted position, this position can is null.
+       */
+      CompletableFuture<Position> getCompactedLastPosition();
+  }
+  ```
+
+* Define a standard CompactionServiceFactory interface to manage `TopicCompactionService`.
+
+  ```java
+  public interface CompactionServiceFactory extends AutoCloseable {
+  
+      CompletableFuture<Void> initialize(PulsarService pulsarService);
+  
+      CompletableFuture<TopicCompactionService> newTopicCompactionService(String topic);
+  }
+  ```
+
+* Implement `PulsarCompactionServiceFactory` and `PulsarCompactionService`
+
+* Combining `CompactedTopicImpl` and `TwoPhaseCompactor` to `PulsarTopicCompactionService`
+
+* Rename `CompactorSubscription` to `PulsarCompactorSubscription`, since it is only applicable to the implementation of Pulsar.
+
+* For CompactorMetrics keep the current implementation, and it only supports `PulsarTopicCompactionService` currently, in the future will use the `Otel` API or other metrics API instead, and customize TopicCompactedService should implement it.
+
+* Fix tests and makes them compatible with new implementations.
+
+## Public-facing Changes
+
+
+### Configuration
+
+broker.conf
+```
+compactionServiceFactoryClassName=org.apache.pulsar.compaction.PulsarCompactionServiceFactory
+```
+
+# Backward & Forward Compatability
+
+## Revert
+
+
+## Upgrade
+
+
+# Alternatives
+
```

Comment:
I think we have alternatives.

- Only make the compactor pluggable
- Make the compaction data serializer and deserializer pluggable

But they are not good for the long-term view of Pulsar

Reply:
Yeah, I add this part